### PR TITLE
Add error preview tooltip to error log row, make exception message property multiline

### LINKED ISSUE #1177: Make viewing exceptions inline better in dashboard
Today the only way to get the entire exception is to hover over the text to see the full thing or to copy it and paste elsewhere:

image

Video:

https://github.com/dotnet/aspire-samples/assets/95136/69029664-2a2f-4ab3-a3c1-25f74088959e

### PR DESCRIPTION
**Pull request by adamint**
Fixes #1177 

Adds a `FluentTooltip` that contains 1) an error preview that includes the exception type and message, and 2) a button through which you can copy that preview. Also adds a property to `PropertyGrid` that allows specific properties to be marked as wrappable (`WrappedProperties`), and adds text wrapping to `ex.Message`

image

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2087)

### CODE REVIEW

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    <FluentIcon Id="@id" Icon="Icons.Filled.Size16.ErrorCircle" Color="Color.Error" Class="severity-icon"/>
+}
+else if (LogEntry.Severity is LogLevel.Warning)
+{
+    <FluentIcon Id="@id" Icon="Icons.Filled.Size16.Warning" Color="Color.Warning" Class="severity-icon"/>
+}
+
+@if (TryGetErrorInformation(out var errorInfo))
+{
+    var test = Guid.NewGuid().ToString();
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor:25]: Better name

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+using System.Diagnostics.CodeAnalysis;
+using Aspire.Dashboard.Otlp.Model;
+using Aspire.Dashboard.Resources;
+using Microsoft.JSInterop;
+
+namespace Aspire.Dashboard.Components;
+
+public partial class LogLevelColumnDisplay
+{
+    private bool TryGetErrorInformation([MaybeNullWhen(false)] out string errorInfo)
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs:14]:
```suggestion
    private bool TryGetErrorInformation([NotNullWhen(true)] out string? errorInfo)
```
Isn't this the idomatic way to do this? I've never seen the opposite.

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+@if (LogEntry.Severity is LogLevel.Error or LogLevel.Critical)
+{
+    <FluentIcon Id="@id" Icon="Icons.Filled.Size16.ErrorCircle" Color="Color.Error" Class="severity-icon"/>
+}
+else if (LogEntry.Severity is LogLevel.Warning)
+{
+    <FluentIcon Id="@id" Icon="Icons.Filled.Size16.Warning" Color="Color.Warning" Class="severity-icon"/>
+}
+
+@if (TryGetErrorInformation(out var errorInfo))
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor:23]: This came up before: exceptions aren't tied to the severity. It's possible to have an exception at any level. The icon shouldn't be used for this.

Also, how would anyone know to hover the cursor over the icon to get exception detail?

File: src/Aspire.Dashboard/Components/Controls/StructuredLogDetails.razor [hunk 4]
```diff
@@ -18,6 +19,7 @@
                       ValueColumnValue="(vm) => vm.Value"
                       NameSort="_nameSort"
                       ValueSort="_valueSort"
-                      HighlightText="@_filter" />
+                      HighlightText="@_filter"
+                      WrappedProperties="@ImmutableArray.Create("ex.Message")"/>
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Controls/StructuredLogDetails.razor:23]: What is the special meaning of `ex.Message`?

These are the standard OTEL properties:
https://github.com/open-telemetry/opentelemetry-dotnet/blob/ebb35248003885c6a3315417d090fb081fccd2b0/src/Shared/SemanticConventions.cs#L96-L100

Can you investigate why you have an exception with non-standard properties.

File: src/Aspire.Dashboard/wwwroot/js/app.js [hunk 5]
```diff
@@ -70,7 +70,7 @@ window.copyTextToClipboard = function (id, text, precopy, postcopy) {
             copyIcon.style.display = 'none';
             checkmarkIcon.style.display = 'inline';
         })
-        .catch(() => {
+        .catch((e) => {
```
Comment (JamesNK) [inline src/Aspire.Dashboard/wwwroot/js/app.js:73]: What is this for? You could write it to the console.log if it's not already written.

File: src/Aspire.Dashboard/Components/Pages/StructuredLogs.razor.css [hunk 6]
```diff
@@ -54,3 +54,7 @@
     grid-area: foot;
     padding-bottom: calc(var(--design-unit) * 2px);
 }
+
+::deep .wrap {
+    text-wrap: wrap;
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/Pages/StructuredLogs.razor.css:59]: This wraps content. But it doesn't respect any of the whitespace (line breaks, indentation) in the content.

For example, this exception message is just one long wrapped line without applying any of the content's whitespace:

image

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        if (LogEntry.Properties.GetValue("ex.Type") is { } exceptionType)
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs:16]: Non-standard property names. Could fallback to them, e.g. look for exception.type first and then ex.Type. Need to figure out why ex.Type is here at all first.

File: src/Aspire.Dashboard/Components/Controls/StructuredLogDetails.razor (hunk 4 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/Components/Controls/StructuredLogDetails.razor:23]: I don't think hardcoding known properties is a good long term situation here. There are many tools that display structured logs out there. What do they do in their UI?

I feel like there should be a way to select a property row and have it switch to multiline, but I don't have another tool that does that off the top of my head as prior art.

**Review by JamesNK — CHANGES_REQUESTED** [review]

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor (hunk 3 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor:23]: What about adding an icon to indicate the log entry has exception data at the end of the message? Where the red box is:

image

Hovering over that shows the error detail popup.

Also, while we're here, there should be a copy button for the message, like there is for the Source column on the resources page.

**Comment by JamesNK** [comment]
Popup displays off the edge of the page:

image

Maximum height is needed.

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+        {
+            var message = LogEntry.Properties.GetValue("ex.Message");
+            var stackTrace = LogEntry.Properties.GetValue("ex.StackTrace");
+            errorInfo = LogEntry.Message + Environment.NewLine + $"{exceptionType}: {message}" + Environment.NewLine + stackTrace;
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs:20]:
```suggestion
            errorInfo = $"{exceptionType}: {message}" + Environment.NewLine + stackTrace;
```
Just include the exception information. There should be a separate way to copy the log message.

Comment (JamesNK) [reply src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs:20]: Actually, from what I see of the data and according to the [spec](https://github.com/open-telemetry/semantic-conventions/blob/45a03e67ce2e109e860c38c60edf6367a1124545/docs/attributes-registry/exception.md), `exception.stacktrace` already includes the exception type and message.

errorInfo should be _just_ `exception.stacktrace`/`ex.StackTrace`.

**Comment by JamesNK** [comment]
I suggest this PR just make changes to the log entry grid. Leave changes to the details view to another PR.

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs (hunk 7 above)
Comment (adamint) [reply src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs:16]: @Erarndt is there a reason these property names are used? They come from here: https://github.com/dotnet/aspire/blob/0282715bfee52b3ed982228a67811b62ec8cb6b3/src/Components/Aspire.RabbitMQ.Client/RabbitMQEventSourceLogForwarder.cs#L156

Comment (adamint) [reply src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs:16]: Falling back to them

File: src/Aspire.Dashboard/Components/Controls/StructuredLogDetails.razor (hunk 4 above)
Comment (adamint) [reply src/Aspire.Dashboard/Components/Controls/StructuredLogDetails.razor:23]: > I feel like there should be a way to select a property row and have it switch to multiline text box that preserves whitespace

I have not checked others, but this does not exist in the msbuild log viewer; it displays as one line.

The solution for now should then just be to wrap all properties. No need to add a toggle

File: src/Aspire.Dashboard/Components/Pages/StructuredLogs.razor.css (hunk 6 above)
Comment (adamint) [reply src/Aspire.Dashboard/Components/Pages/StructuredLogs.razor.css:59]: responded [here](https://github.com/dotnet/aspire/pull/2087#discussion_r1480422514)

**Comment by adamint** [comment]
They're both related changes, as part of the same experience; IMO should stay in the same PR

**Comment by adamint** [comment]
Made the popup slightly wider, added max height with scrolling.

image

Preserves whitespace in property grid
image

File: src/Aspire.Dashboard/Components/Controls/StructuredLogDetails.razor (hunk 4 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/Components/Controls/StructuredLogDetails.razor]: msbuild log viewer is probably made for devs, by devs. We should look at what other tools in this space do. What does AppInsights in azure do? Or datadog, seq, etc.

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs (hunk 7 above)
Comment (JamesNK) [reply src/Aspire.Dashboard/Components/ResourcesGridColumns/LogLevelColumnDisplay.razor.cs]: Create an issue for this.

**Comment by JamesNK** [comment]
> Made the popup slightly wider, added max height with scrolling. I kept the icon on the left of the column, as it conforms with other columns.
> 
> image

Ok. I'd prefer it if the icon color was white (or whatever the font color is). A logged exception isn't necessarily a bad thing.

I think the tooltip needs a title, e.g. "<h3>Exception details</h3>"

> 
> Preserves whitespace in property grid image

It looks ok. When preserving whitespace, you need to be careful about what whitespace gets accidentally added in the HTML from identing tags.

Remove this whitespace here:
image

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/LogMessageColumnDisplay.razor.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+        {
+            errorInfo = null;
+            return false;
+        }
+
+        if (string.IsNullOrEmpty(stackTrace))
+        {
+            var message = GetProperty("exception.message") ?? GetProperty("ex.Message");
+            var type = GetProperty("exception.type") ?? GetProperty("ex.Type");
+            errorInfo = LogEntry.Message + Environment.NewLine + $"{type}: {message}" + Environment.NewLine + stackTrace;
```
Comment (JamesNK) [inline src/Aspire.Dashboard/Components/ResourcesGridColumns/LogMessageColumnDisplay.razor.cs:26]:
```suggestion
            errorInfo = $"{type}: {message}" + Environment.NewLine + stackTrace;
```
Exception details doesn't include log enetry message.

File: src/Aspire.Dashboard/Components/Controls/StructuredLogDetails.razor (hunk 4 above)
Comment (davidfowl) [reply src/Aspire.Dashboard/Components/Controls/StructuredLogDetails.razor]: There are others that we will need to be expandable. Can it be based on the content truncation?

Comment (adamint) [reply src/Aspire.Dashboard/Components/Controls/StructuredLogDetails.razor]: > Can it be based on the content truncation

possibly, but I don't know how to do this

File: src/Aspire.Dashboard/Components/ResourcesGridColumns/LogMessageColumnDisplay.razor [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+                Id="@iconId"
+                Icon="Icons.Filled.Size16.DocumentError"
+                Color="Color.Warning"
+                Class="severity-icon"/>
+
+    var copyButtonId = Guid.NewGuid().ToString();
+
+    <FluentTooltip Anchor="@iconId" MaxWidth="650px">
+        <div style="padding: 10px">
+            <pre style="white-space: pre-wrap; max-height: 400px; overflow: auto;">@errorInfo</pre>
```
[the remark this answers is not in the document]
Comment (adamint) [reply src/Aspire.Dashboard/Components/ResourcesGridColumns/LogMessageColumnDisplay.razor:24]: changed

**Comment by adamint** [comment]
image
image
Added title, removed the extra spacing that you pointed out James. I don't think the icon should be the same color as the text, as it will not stand out, but how is accent?

**Comment by JamesNK** [comment]
> but how is accent?

Yeah that's good.

**Review by JamesNK — APPROVED** [review]
👍 

An odd thing I see is the scroll position is maintained between the tooltip closing and reopened. It feels like a bug to me but maybe some people would want that? IDK

**Comment by adamint** [comment]
> 👍
> 
> An odd thing I see is the scroll position is maintained between the tooltip closing and reopened. It feels like a bug to me but maybe some people would want that? IDK

That seems like a positive to me, but I don't know. Happy to change if we get negative feedback on it

**Comment by dersia** [comment]
thanks for the copy to clipboard button!

**Pull request merged**